vendor/noise_handshake/src/handshake.rs

```
@@ -0,0 +1,217 @@
+        match self {
+            Self {
+                noise: Some(ref mut noise),
+                io: Some(_),
+                next:
+                    Step::Write {
+                        ref mut buffer,
+                        ref mut len,
+                        ..
+                    },
+            } if *len == 0 => {
+                *len = noise
+                    .write_message(&[], buffer)
+                    .expect("Cannot encode the message");
+                self.poll()
+            }
+            Self {
+                noise: Some(ref mut noise),
+                io: Some(ref mut io),
+                next:
+                    Step::Write {
+                        buffer,
+                        ref mut written_bytes,
+                        len,
+                    },
+                ..
+            } => {
+                while written_bytes < len {
+                    *written_bytes = try_ready!(io.poll_write(&buffer[*written_bytes..*len]));
+                }
+                if noise.is_handshake_finished() {
+                    Ok(Async::Ready(self.wrap_up()))
+                } else {
+                    self.next = Step::read();
+                    self.poll()
+                }
+            }
+            Self {
+                noise: Some(ref mut noise),
+                io: Some(ref mut io),
+                next:
+                    Step::Read {
+                        mut enc_buffer,
+                        ref mut len,
+                    },
+            } => {
+                let mut dec_buffer = [0u8; NOISE_MAX_SIZE];
+
+                *len += try_ready!(io.poll_read(&mut enc_buffer[*len..]));
+                if *len > 0 {
+                    match noise.read_message(&enc_buffer[..*len], &mut dec_buffer) {
+                        Ok(_) => self.next = Step::write(),
+                        Err(e) => debug!("Error decoding message: {:?}", e),
```

Knowing the size of what we are expecting would remove this weird Err debug thing. If we know what we are looking for we can just return a proper error.